Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+    /// Use initialized service within the handler
+    try await postgresClient.query(...)
+}
+
+/// Use ServiceLifecycle to manage the initialization and termination
+/// of the services as well as the LambdaRuntime
+let serviceGroup = ServiceGroup(
+    services: [postgresClient, runtime],
+    configuration: .init(gracefulShutdownSignals: [.sigterm]),
+    logger: logger
+)
+try await serviceGroup.run()
+```
+
+#### Simplifying Codable support
+
+A detailed explanation is provided in the Codable Support section. In short, much of the boilerplate code defined for each handler protocol in `Lambda+Codable` and `Lambda+String` will be replaced with a single `LambdaCodableAdapter` class.
+
+This adapter class is generic over (1) any handler conforming to `LambdaHandler`, (2) the user-specified input and output types, and (3) any decoder and encoder conforming to `LambdaEventDecoder` and `LambdaOutputDecoder`. The adapter will wrap the underlying handler with encoding/decoding logic.
+
+#### Simplifying the handler protocols
+
+There are four different handler protocols in the current API (`SimpleLambdaHandler`, `LambdaHandler`, `EventLoopLambdaHandler`, `ByteBufferLambdaHandler`). As noted in the **Current Limitations** section, the ease-of-use varies for each handler protocol and users may not be able to easily determine which protocol best serves their use-case without spending time digging into the library. To reduce this problem and provide users with clear-cut options, we propose replacing all of the existing handler protocols with just two: `LambdaHandler` and `StreamingLambdaHandler`. Both will be explained in the **Detailed Solution** section.
+
+## Detailed Solution
+
+Below you’ll find explanation’s for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below), which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
+    /// Write a response part into the stream
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type `ByteBuffer`) will be removed.
+
+A new function `backgroundTask()` will also be added. This will allow tasks to be run in the background while and after the response is/has been sent. Please note that `LambdaContext` will not be Sendable anymore.
```

Comment:
In order to allow adding background tasks in a structured way, we will need to back the `LambdaContext` with a `TaskGroup`. Since `TaskGroup` is not `Sendable` we can't mark `LambdaContext` as `Sendable`. However if users need any property (all of them are `Sendable`) out of the `LambdaContext` they can just get that property out of it and then pass it around or close over it.

Reply:
Why does the  context need to hold the `TaskGroup`? I smell structured concurrency violations :)